patch sequence page (#1205)
ocaml/v3.ocaml.org: data/tutorials/ds_05_seq.md

category: "data-structures"
date: 2023-01-12T09:00:00-01:00
---

# Sequences

## Prerequisites

You should be comfortable with writing functions over lists and options.

## Introduction

Sequences are very much like lists. However from a pragmatic perspective, one
should imagine they may be infinite. That's the key intuition to understanding
and using sequences.

One way to look at a value of type `'a Seq.t` is to consider it as a list, with
a twist when it's not empty: a frozen tail. To understand this analogy,
consider how sequences are defined in the standard library:
```ocaml
type 'a node =
  | Nil
  | Cons of 'a * 'a t
and 'a t = unit -> 'a node
```
This is the mutually recursive definition of two types: `Seq.node`, which is
almost the same as `list`:
```ocaml
type 'a list =
  | []
  | (::) of 'a * 'a list
```
and `Seq.t`, which is merely a type alias for `unit -> 'a Seq.node`. The whole
point of this definition is the second argument's type `Seq.Cons`, which is a
function returning a sequence while its `list` counterpart returns a list. Let's
compare the constructors of `list` and `Seq.node`:
1. Empty lists and sequences are defined the same way, a constructor without any

>    parameter: `Seq.Nil` and `[]`.

Reviewer: ```suggestion
   parameters: `Seq.Nil` and `[]`.
```

In OCaml, any value `a` of type `t` can be turned into a constant function by
writing `fun _ -> a`, which has type `'a -> t`. When writing `fun () -> a`
instead, we get a function of type `unit -> t`. Such a function is called a
[_thunk_](https://en.wikipedia.org/wiki/Thunk). Using this terminology, `Seq.t`
values are thunks. With the analogy used earlier, `a` is frozen in its thunk.

Here is how to build seemingly infinite sequences of integers:
```ocaml
# let rec ints n : int Seq.t = fun () -> Seq.Cons (n, ints (n + 1))
val ints : int -> int Seq.t = <fun>
```
The function `ints n` looks as if building the infinite sequence `(n; n + 1; n +
2; n + 3;...)`. In reality, since there isn't an infinite amount of distinct
values of type `int`, those sequences aren't indefinitely increasing. When
reaching `max_int`, the values will circle down to `min_int`. They are
ultimately periodic.

The OCaml standard library contains a module on sequences called
[`Seq`](/releases/5.0/api/Seq.html). It contains a `Seq.iter` function, which
has the same behaviour as `List.iter`. Writing this:
```ocaml
# Seq.iter print_int (ints 0);;
```

> in an OCaml top-level means “print integers forever,” and you have to press

Reviewer: ```suggestion
in an OCaml toplevel means “print integers forever,” and you have to press
```

the [Sieve of
Eratosthenes](https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes). Here it is:
```ocaml
let rec sieve seq () = match seq () with
  | Seq.Cons (m, seq) -> Seq.Cons (m, sieve (Seq.filter (fun n -> n mod m > 0) seq))
  | seq -> seq
let facts = ints_from 2 |> sieve;;
```

This code can be used to generate lists of prime numbers. For instance, here is
the list of 100 first prime numbers:
```ocaml
# facts |> take 100 |> List.of_seq;;
- : int list =
[2; 3; 5; 7; 11; 13; 17; 19; 23; 29; 31; 37; 41; 43; 47; 53; 59; 61; 67; 71;
 73; 79; 83; 89; 97; 101; 103; 107; 109; 113; 127; 131; 137; 139; 149; 151;
 157; 163; 167; 173; 179; 181; 191; 193; 197; 199; 211; 223; 227; 229; 233;
 239; 241; 251; 257; 263; 269; 271; 277; 281; 283; 293; 307; 311; 313; 317;
 331; 337; 347; 349; 353; 359; 367; 373; 379; 383; 389; 397; 401; 409; 419;
 421; 431; 433; 439; 443; 449; 457; 461; 463; 467; 479; 487; 491; 499; 503;
 509; 521; 523]
```

The function `sieve` is recursive in OCaml and common sense. It is defined using
the `rec` keyword and calls itself. However, some call that kind of function
[_corecursive_](https://en.wikipedia.org/wiki/Corecursion). This word is used to

> emphasize that, by design, it does not terminate. Strictly speaking, the sieve

Reviewer: ```suggestion
emphasise that, by design, it does not terminate. Strictly speaking, the sieve
```

val fibs : int -> int -> int Seq.t = <fun>
```
It can be used to produce some Fibonacci numbers:
```ocaml
# fibs 0 1 |> Seq.take 10 |> List.of_seq;;
- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
```
Why is it so? The key difference lies in the recursive call `fibs n (n + m)`. In
the former definition, the application is complete because `fibs` is provided
all the arguments it expects. In the latter definition, the application is
partial because the `()` argument is missing. Since evaluation is
[eager](https://en.wikipedia.org/wiki/Evaluation_strategy#Eager_evaluation) in
OCaml, in the former case, evaluation of the recursive call is triggered and a
non-terminating looping occurs. In contrast, in the latter case, the partially
applied function is immediately returned as a
[closure](https://en.wikipedia.org/wiki/Closure_(computer_programming)).

Sequences are functions, as stated by their type:
```ocaml
# #show Seq.t;;
type 'a t = unit -> 'a Seq.node
```
Functions working with sequences must be written accordingly.
* Sequence consumer: partially applied function parameter
* Sequence producer: partially applied function result

> When code dealing with sequences does not behave as expected like if it is

Reviewer: ```suggestion
When code dealing with sequences does not behave as expected, like if it is
```

crashing or hanging, there's a fair chance a mistake like in the first
definition of `fibs` was made.

## Sequences for Conversions

Throughout the standard library, sequences are used as a bridge to perform
conversions between many datatypes. For instance, here are the signatures of
some of those functions:
* Lists
  ```ocaml
  val List.of_seq : 'a list -> 'a Seq.t
  val List.to_seq : 'a Seq.t -> 'a list
  ```
* Arrays
  ```ocaml
  val Array.of_seq : 'a array -> 'a Seq.t
  val Array.to_seq : 'a Seq.t -> 'a array
  ```
* Strings
  ```ocaml
  val String.of_seq : string -> char Seq.t
  val String.to_seq : char Seq.t -> string
  ```

> Similar functions are also provided for sets, maps, hash tables (`Hashtbl`) and

Reviewer: ```suggestion
Similar functions are also provided for sets, maps, hash tables, (`Hashtbl`) and
```
